Jinja2 template system

jkfran opened the pull request:
# Description

This PR will change completely the structure of our deployments configs, and we will no longer save Kubernetes objects directly in this repo.

Instead, we will have small files with variables values per each project, and we will generate the Kubernetes output when needed.

## List of changes
- Remove all Kubernetes manifests.
- Jinja templates under `templates` folder.
- All projects config values under  `sites` folder.
- konf.py script to render the templates per project
- qa-templates-checker.sh to make sure we are not breaking the current Kubernetes objects
- Updated qa-deploy to test projects locally
- Updated README.md

## New requirements

We have to be sure that the following system dependencies are installed:
- python3-jinja2
- python3-yaml

## How it works

Take a look to the updated [README.md](README.md) file to understand how konf.py works.

# QA
Download this [gist](https://gist.github.com/jkfran/fa5742f73e4b590b7ca65bb308fc939c) and place it on the root of this branch. `qa-templates-checker.sh` is a script that I created in order to verify that this template system will not produce any unwanted changes to our current configs. Feel free to take a look into it before using it.

## How the QA script works

Basically we are using `kubectl diff` to compare against all our configuration in the git branch "master". We first load all the Kubernetes objects in our system and then we do `kubectl diff` for each project.

## QA Instructions

- Make sure you have the latest version of the master branch (It's important because we do the diff using this branch)
- Do a git checkout to this branch
- run `./qa-templates-checker.sh`
- It will output all projects found and if there are changes detected or not
- Look the output and the diff.log file

## Interpret the output

We can see if there are changes by the environment (production or staging) or missing projects.
If there are changes a new file called `diff.log` has been generated where you can see these changes.

The `generation` property in the `diff.log` file can be ignored, it's going to appear always that we made a change in any Kubernetes object:
```
-  generation: 1
+  generation: 2
```

It doesn't make sense to keep this QA script in the repository after the merge but it has been necessary for the development and also to validate this PR

# Next steps

When the approval of this PR occurs, it will be necessary to start working on the necessary changes in Jenkins

nottrobin commented on sites/canonical-com.yaml:
```diff
@@ -0,0 +1,28 @@
+container:
+  image: prod-comms.docker-registry.canonical.com/canonical.com
+
+useSquid: true
```
This means using the `squid.internal` proxy settings, right?

This should really be the default for all sites, and do I think ideally it wouldn't have a setting - it would just be applied. The only thing is, if we change this to make it default for all sites then we have to be careful to roll out these changes one site at a time to make sure they don't break anything. Still I think that's worth doing.

Would you mind removing this option and simply making it so these settings are applied for all sites?

jkfran replied:
Renamed to useProxy as we agreed.

nottrobin replied:
Filed https://github.com/canonical-web-and-design/deployment-configs/issues/280 for removing this option, after this is merged.

jkfran replied:
Thank you Robin!

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+    tag: a264efb326485
+```
+
+## wwwPrefixHost
````
Rather than making this a choice, could we do logic to say that if it's got only 1 dot in it then we include www?

Basically, all apex domains (ubuntu.com, cloud-init.io) should have www, and all non-apex ones (design.ubuntu.com) shouldn't. We can enforce this pattern better through logic.

jkfran replied:
Removed

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+tlsSecretName: cloud-init-org-tls
+```
+
+## useSquid
````
As mentioned, could we remove this one and simply make all sites have the proxy settings?

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+useSquid: True
+```
+
+## readinessPath
````
Again, could we remove this option and make it always use `/status/_check`? Then if this doesn't work for any site we have, we should just fix the sites to make sure it does work. It should be pretty trivial to do this in all cases.

jkfran replied:
I have left it for now as we agreed

nottrobin replied:
I filed https://github.com/canonical-web-and-design/deployment-configs/issues/281 for removing this option down the line.

jkfran replied:
Perfect!

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+readinessPath: /
+```
+
+## readinessPeriodSeconds
````
I don't think these should be settings - we should just use the settings we choose across the board.

jkfran replied:
Removed

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+memory: 512Mi
+```
+
+## envSecrets
````
This should just be called `env`, or `environment_variables`. These are explicitly *not* secrets, as their values are included in this repository in plain text for anyone to read.

jkfran replied:
Renamed to env

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+    - docs.staging.jujucharms.com
+```
+
+### tlsSecret
````
We want the name of the TLS secrets to be absolutely consistent, and to be derived from the domain. For `ubuntu.com`, it's `ubuntu-com-tls`. For this reason, I again don't think this should be a setting. It should be computed from the domain name.

jkfran replied:
Removed

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+  noTls: true
+```
+
+## nginxConfig
````
Could we include the word `snippet` in here somewhere? To match the k8s wording - https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#server-snippet - and indicate that it's not a full nginx config file.

jkfran replied:
Renamed to `nginxConfigurationSnippet`

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+readinessPeriodSeconds: 10
+```
+
+## readinessSuccessThreshold
````
Likewise

jkfran replied:
Removed

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+readinessSuccessThreshold: 5
+```
+
+## readinessTimeoutSeconds
````
Likewise

jkfran replied:
Removed

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+readinessTimeoutSeconds: 5
+```
+
+## memory
````
This one we should keep =)

jkfran replied:
=) renamed to memoryLimit

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+wwwPrefixHost: false
+```
+
+## tlsSecretName
````
What's the difference between this and `tlsSecret`?

jkfran replied:
Removed

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+    image: prod-comms.docker-registry.canonical.com/canonical.com
+```
+
+### tag
````
Will we ever define the tag here? I think this always gets dynamically assigned at deploy time doesn't it? So it's probably not a necessary setting here.

jkfran replied:
Removed

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+noIngress: true
+```
+
+## container
````
Why keep `image` under `container`? Wouldn't it be more straightforward to simply put an `image` setting at the top level?

jkfran replied:
Removed, now `image` is at the top level

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+domain: canonical.com
+```
+
+## noIngress
````
Are there any cases where there's no ingress? I think we should explore the cases where you're using this - I think what's going on in those cases could be described a bit better.

jkfran replied:
Removed because of the new `routes` property

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+Example:
+```
+extraHosts:
+  - host:
````
Why the secondary `- host` key? Why can't it simply be:

``` yaml
extraHosts:
  - example.com
```

?

jkfran replied:
Updated, now it is like:

```
extraHosts:
  - domain: example.com
```

nottrobin commented on sites/README.md:
````diff
@@ -0,0 +1,293 @@
+```
+- host:
+  - ubuntu.net
+  noTls: true
````
Is this yaml correct? Shouldn't it be more like:

``` yaml
- host:
  - ubuntu.net:
      noTls: true
```

Or something?

jkfran replied:
Updated, now it is like:

```
extraHosts:
  - domain: mirserver.io
    useParentTLS: True
```

jkfran commented:
Hi @nottrobin ,

I have made the changes we agreed. It would be great if you could check this again.

I have seen that usn.ubuntu.com have this customization to stop de container:
https://github.com/canonical-web-and-design/deployment-configs/blob/e2758acc3f0f80e7d397cd5368ecf793817e30d4/services/usn-ubuntu-com.yaml#L40

And it's going to disappear with the templates, should we keep it?

Also, I have to add the property `containerPort` because the docker for IRC bot is running on the port 8080 instead of 80 like the other projects:
https://github.com/canonical-web-and-design/deployment-configs/blob/e2758acc3f0f80e7d397cd5368ecf793817e30d4/services/webteam-ircbot-canonical-com.yaml#L34

Let me know what do you think about this. Thank you.

nottrobin commented:
@jkfran I've filed https://github.com/canonical-web-and-design/webteam-hubot/issues/28 to fix the `8080` container port, but for now I suppose we have to offer an option to override the default container port :(. Let me know when you've done that.

I've also filed https://github.com/canonical-web-and-design/deployment-configs/issues/282 about updating the `STOPSIGNAL` in relevant images. I think it's fine to remove that line from usn.ubuntu.com config for now.

nottrobin commented on README.md:
````diff
@@ -2,51 +2,58 @@
-### `ingresses`
+It is possible to generate only some of these objects, E.g.:
+``` bash
+./konf.py staging sites/canonical-com.yaml --templates ingress
````
Is this still possible? We discussed changing this.

jkfran replied:
Oh, I didn't update this README.md, I will update it soon

nottrobin replied:
No rush. It's gonna take me a while to properly review this.

nottrobin commented on create-project:
```diff
@@ -1,79 +1,70 @@
-            'sentry_dsn': sentry_dsn
-        }
-    )
+import yaml
```
This project is now going to need a `requirements.txt` with at least `pyyaml` in it.

nottrobin commented on konf.py:
```diff
@@ -0,0 +1,139 @@
+#! /usr/bin/env python3
+
+import yaml
+import jinja2
```
This is also going to need to be in `requirements.txt`.

jkfran replied:
Okay, can it be confusing for Jenkins? We agreed with IS for security reason to use the Debian packages there.

jkfran replied:
What if I make this a snap? I mean, the `konf.py` and the Jinja2 `templates` folder, this would also help us with what we talk about having each file in its own repository.

nottrobin replied:
As discussed, I don't think we should publish this config generator as a package.

This is because, firstly, we will only ever use it for one thing - for applying k8s configs from our Jenkins installation - so packaging is not any use for general distribution.

Secondy, when we make changes to these configs, we'll likely want to coordinate the release of those changes with a change to one of our projects, and releasing those changes out of sync could potentially be catastrophic.

If we keep the configs in this repo, then our release process looks like:

1. push these changes to master
2. run the Jenkins job which will pull down the latest version from master and run it

But if we were to publish it as a package, then our release flow gets more complicated:

1. push these changes to master
2. publish the package
3. update the version of that package on Jenkins
4. run the Jenkins job

I'm worried that this increases the cost of step getting forgotten, possibly to catastrophic effect, especially when I don't see an obvious benefit to packaging it up.

Thirdly, if we package it up then this repo will get more complicated and harder to understand, as it needs to contain packaging infrastructure.

nottrobin commented on templates/deployment.yaml:
```diff
@@ -0,0 +1,68 @@
+              containerPort: {{ data.containerPort | default(80) }}
+
+          envFrom:
+            {%- if data.useProxy != False %}
```
Actually, I just thought you might as well add all these environment variables directly to the `env:` block, but keep this logical wrapper.

jkfran replied:
Done :)

nottrobin commented on templates/ingress.yaml:
```diff
@@ -0,0 +1,76 @@
+    {%- endif %}
+
+spec:
+  {%- if not data.noTls and data.noTls is not defined %}
```
Why do you need both `not` and `not defined`?

jkfran replied:
No, you are right I don't need both

nottrobin commented on konf.py:
```diff
@@ -0,0 +1,139 @@
+
+# Custom Jinja2 functions
+@jinja2.contextfunction
+def getSiteName(context, addStaginPrefix=False):
```
In Python, function and variable names should be written in snake case:

> Function names should be lowercase, with words separated by underscores as necessary to improve readability.
> 
> Variable names follow the same convention as function names.
> 
> mixedCase is allowed only in contexts where that's already the prevailing style (e.g. threading.py), to retain backwards compatibility.

https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

(You'd do well to read [PEP8](https://www.python.org/dev/peps/pep-0008/) and [PEP257](https://www.python.org/dev/peps/pep-0257/) in full if you haven't already)

nottrobin replied:
Following on from this - we tend to use [black](https://pypi.org/project/black/) with `--line-length 79` to format all our code - did you run it over your code here?

It's not too hard to configure VSCode or any other editor to do this for you every time you save or whatever. Ask if you want help with that.

nottrobin replied:
There's also a `g` missing in `addStaginPrefix`

nottrobin replied:
I've create this issue - but don't worry about it for this PR, we can solve it down the line.

https://github.com/canonical-web-and-design/deployment-configs/issues/283

jkfran replied:
Sorry, I changed it

nottrobin commented on konf.py:
```diff
@@ -0,0 +1,139 @@
+
+
+@jinja2.contextfilter
+def siteDomain(context, s):
```
I'm not a huge fan of single-character variable names

jkfran replied:
Okay!

nottrobin commented on konf.py:
```diff
@@ -0,0 +1,139 @@
+        templateLoader = jinja2.FileSystemLoader("./templates")
+        self.tplEnv = jinja2.Environment(loader=templateLoader)
+
+        # Add custom jinja2 functions and filters
```
As far as I can tell you only call `self.tplEnv.render` in one place. Did you add these functions here in the global space for performance or something? Because they could be passed at render time couldn't they?

nottrobin commented on qa-templates-checker.sh:
```diff
@@ -0,0 +1,64 @@
+#! /usr/bin/env bash
```
Please remove this from the repo, place in a [gist](https://gist.github.com/), and add the link to the QA steps.

jkfran replied:
Done

nottrobin commented on sites/README.md:
```diff
@@ -0,0 +1,236 @@
+
+**Type:** String
+
+**Default:** /_status/check
```
I think, you at least make the GitHub syntax highlighter happy, you should escape the underscore:

```suggestion
**Default:** /\_status/check
```

nottrobin replied:
Or maybe putting it in backticks would also work?

```suggestion
**Default:** `/_status/check`
```

In which case you should do that with all the properties =)

jkfran replied:
Done

nottrobin commented on README.md:
```diff
@@ -2,51 +2,56 @@
+When creating a new project, you can either copy an existing projects values file and modify the file as you want. Here is a document explaining all the possibles values:
 
-Or, use `./create-project`. To auto generate the files for a new project. Run `./create-project project.name` and the tool will create the config files for you. 
+- [Values file manual](sites/README.md)
```
I love that you made this :+1: :rocket:

jkfran replied:
Haha thank you!

jkfran replied:
Are you trying to keep motivated? :P

nottrobin commented:
@jkfran also could you rebase this with master? GitHub says it has a conflict.

jkfran commented:
@nottrobin Done, but should I rebase to make all these changes of this just one commit?

Thanks Robin, I think we're closer!

nottrobin commented:
Issues I've found in the diff:

## staging-api

You're creating `staging-api.staging.snapcraft.io`, which isn't needed - `staging-api.snapcraft.io` should already  be in the `staging` namespace.

## ubuntu.com/blog

The generated `ubuntu-com` ingress config looks like this:

``` yaml
  rules:                                                                                                                                                                                                           
  - host: ubuntu.com                                                                                                                                                                                               
    http: &http_service                                                                                                                                                                                            
      paths:                                                                                                                                                                                                       
      - path: /                                                                                                                                                                                                    
        backend:                                                                                                                                                                                                   
          serviceName: ubuntu-com                                                                                                                                                                                  
          servicePort: 80                                                                                                                                                                                          
      - path: /blog                                                                                                                                                                                                
        backend:                                                                                                                                                                                                   
          serviceName: ubuntu-com-blog                                                                                                                                                                             
          servicePort: 80                                                                                                                                                                                          
```

Whereas the current one looks like this:

``` yaml
  rules:
  - host: ubuntu.com
    http: &ubuntu-com_service
      paths:
      - path: /blog
        backend:
          serviceName: ubuntu-com-blog
          servicePort: 80
      - path: /
        backend:
          serviceName: ubuntu-com
          servicePort: 80
```

I believe the order of these paths is actually very important - it only works if `/blog` is before `/`. Could we preserve this, and make sure that all sub-paths appear before the root in all configs?

## ubuntu.com SSL redirect

I forget what we discussed about this, but if we're to actually release the production configs this side of Christmas, we probably need to keep the `nginx.ingress.kubernetes.io/ssl-redirect: "false"` on the `ubuntu-com` ingress.

## Extra 360 staging service

For some reason, just for 360, we seem to be creating a `threesixty-staging-canonical-com` service:

``` diff
--- /tmp/LIVE-804280665/v1.Service.default.threesixty-staging-canonical-com	2019-11-20 14:29:32.651462463 +0000
+++ /tmp/MERGED-686682852/v1.Service.default.threesixty-staging-canonical-com	2019-11-20 14:29:32.651462463 +0000
@@ -0,0 +1,20 @@
+apiVersion: v1
+kind: Service
+metadata:
+  creationTimestamp: "2019-11-20T14:29:32Z"
+  name: threesixty-staging-canonical-com
+  namespace: default
+  selfLink: /api/v1/namespaces/default/services/threesixty-staging-canonical-com
+  uid: e79413ee-e6a6-493b-bd00-a945b57f4509
+spec:
+  ports:
+  - name: http
+    port: 80
+    protocol: TCP
+    targetPort: http
+  selector:
+    app: 360.canonical.com
+  sessionAffinity: None
+  type: ClusterIP
+status:
+  loadBalancer: {}
```

This should just be called `threesixty-canonical-com`, like with all the other services.

nottrobin commented:
The next test I'll do is to actually try deploying all the sites with these new configs on a local microk8s. But I'll do that after we've resolved the issues in my previous comment.

jkfran commented:
Hey @nottrobin

## staging-api

For this site we are currently not using two environments, we just have staging so I have made some modifications to allow to specify the staging domain instead of generating a new one. The `qa-templates-checker.sh` runs for production and staging, but we will skip production on Jenkins, right?

## ubuntu.com/blog

I didn't know about the importance of the path order. I'm okay changing the order, but what about:
https://github.com/canonical-web-and-design/deployment-configs/blob/master/ingresses/production/docs-ubuntu-com.yaml#L21-L36

Looks like I will change the output for docs.ubuntu.com if I put the `/` path the last one.

## ubuntu.com SSL redirect

I thought we had agreed to remove it. Okay, I just added a new property: `nginxSSLRedirect`.

## Extra 360 staging service

I fixed it, it was only 360 because it had a name assigned since the name can't be generated with the domain.

nottrobin commented:
@jkfran thanks for all that.

> ubuntu.com SSL redirect
> 
> I thought we had agreed to remove it.

Yeah, I remember that. I think my thinking then was that we would be moved to the content-cache pretty soon which will support HTTPS. But actually we should probably merge this one sooner than that.

nottrobin commented:
@jkfran when I do `./qa-deploy production sites/ubuntu-com.yaml`, I end up with 1 replica of ubuntu-com, but 5 of ubuntu-com-blog:

```
NAME                                   READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/default-http-backend   1/1     1            1           2m11s
deployment.apps/ubuntu-com             1/1     1            1           100s
deployment.apps/ubuntu-com-blog        5/5     5            5           100s
```

nottrobin commented on sites/canonical-com.yaml:
```diff
@@ -0,0 +1,24 @@
+domain: canonical.com
```
Sorry, I should have caught this much earlier. Could we please name these files by the domain name - as in with dots? Like `sites/canonical.com.yaml`?

'cos now they're not a file for a specific Kubernetes object (where the naming has hyphens) but instead a file for the whole config for a site.

jkfran replied:
Done!

nottrobin commented:
I believe you can now delete `configmaps/global-envvars.yaml` and `configmaps/proxy-config.yaml`.

nottrobin commented on sites/README.md:
```diff
@@ -0,0 +1,249 @@
+
+**Default:** `True`
+
+This will use our ConfigMap object "proxy-config", you can take a look to this object here: `configmaps/proxy-config.yaml`
```
This is no longer accurate is it?

jkfran replied:
No, you are right. I updated the description.

nottrobin commented:
@jkfran okay I tested this on microk8s with both staging and production configs for `ubuntu.com` and `docs.ubuntu.com`. They both work well, apart from the issue with too many replicas for the subpaths. I think we're very close.

Once you've addressed my points above I think we can merge this, and then we just have to very cautiously roll it out next week, check each site as we go.

nottrobin commented on sites/juju.is.yaml:
```diff
@@ -0,0 +1,23 @@
+
+image: prod-comms.docker-registry.canonical.com/juju.is
+
+globalEnvVars: false
```
Is this used?
